File: labosi/Primjeri/Problem misionara i kanibala/misionari.hs
-- |
-- Sveučilište u Zagrebu, Fakultet elektrotehnike i računarstva
-- Umjetna inteligencija
--
-- Pretraživanje prostora stanja: Problem misionara i kanibala
-- v1.0
--
-- Copyright: (c) 2011 Jan Snajder <jan.snajder@fer.hr>
--
-- Zadatak:
-- Napišite program koji ce pretraživanjem u širinu pronaći optimalno
-- rješenje problema misionara i kanibala. Problem je opisan na sljedeći 
-- način: tri misionara i tri kanibala potrebno je jednim camcem prevesti s 
-- jedne strane obale rijeke na drugu, pri cemu se niti u jednom trenutku na
-- jednoj strani obale ne smije naći više kanibala nego misionara. Čamac 
-- može prevesti najviše dvije osobe i ne može ploviti prazan. Optimalno
-- rješenje je ono s najmanjim brojem koraka. Program treba ispisati
-- rješenje u obliku niza operatora i stanja
-- koja dovode do ciljnog stanja.
-- Primjer: http://www.learn4good.com/games/puzzle/boat.htm
--
------------------------------------------------------------------------------

import Data.List
import Control.Monad
import qualified Data.Set as S

------------------------------------------------------------------------------
-- Definicija problema
------------------------------------------------------------------------------

-- Pozicija čamca (lijeva ili desna obala).
data Boat = L | R 
  deriving (Eq,Show,Ord)

-- Stanje je prikazano kao trojka sa sljedećim elementima: broj misionara 
-- na lijevoj obali, broj kanibala na lijevoj obali i pozicija čamca.
type State = (Int,Int,Boat)

-- Početno stanje (tri misionara i tri kanibala su na lijevoj obali).
initState :: State
initState = (3,3,L)

-- Ispitni predikat; vraća True za ciljno stanje.
isGoal :: State -> Bool
isGoal (0,0,R) = True
isGoal _       = False

-- Funkcija za generiranje valjanih sljedbenika zadanog stanja.
nextState :: State -> [State]
nextState (m,c,b) = do 
  (x,y) <- [(1,1),(2,0),(0,2),(1,0),(0,1)]
  let s = if b==L then (m-x,c-y,R) else (m+x,c+y,L)
  guard $ safe s
  return s
  where safe (m,c,_) = (m == 0 || m >= c) && (m == 3 || m <= c)

------------------------------------------------------------------------------
-- Pretraživanje u širinu
------------------------------------------------------------------------------
  
-- Pretraživanje u širinu s pamćenjem posjećenih stanja. Funkcija
-- uzima početno stanje, funkciju sljedbenika i ispitni predikat, a
-- vraća listu stanja od početnog do ciljnog stanja, odnosno praznu
-- listu ako rješenje nije pronađeno.
bfSearch :: (Ord a) => a -> (a -> [a]) -> (a -> Bool) -> [a]
bfSearch s0 succ goal = step [(s0,[])] S.empty
  where step [] v = []
        step (x@(s,ss):xs) v 
          | goal s       = reverse $ s:ss
          | S.member s v = step xs v
          | otherwise    = step (xs ++ expand x) (S.insert s v)
        expand (s,ss) = [(t,s:ss) | t <- succ s]

-- Npr.
-- > bfSearch initState nextState isGoal

-- Pomoćna funkcija za prikaz operatora (prijelaza između dva stanja).
showStep :: State -> State -> String
showStep s1@(m1,c1,b1) s2@(m2,c2,_) = concat $ case b1 of
  L -> [showMC (m1-x,c1-y), " (", showMC (x,y), ") ---> ", showMC (3-m1,3-c1)]
  R -> [showMC (m2-x,c2-y), " <--- (", showMC (x,y), ") ", showMC (3-m2,3-c2)]
  where showMC (m,c) = show m ++ "M " ++ show c ++ "C"
        (x,y) = (abs $ m1-m2,abs $ c1-c2)
        space n = replicate n ' '

solve :: IO ()
solve = putStr . unlines $ zipWith showStep ss (tail ss)
  where ss = bfSearch initState nextState isGoal

File: labosi/Primjeri/Problem misionara i kanibala/misionari.cpp
/* Sveučilište u Zagrebu, Fakultet elektrotehnike i računarstva
   Umjetna inteligencija
 
   Pretraživanje prostora stanja: Problem misionara i kanibala
   v1.0
 
   Copyright: (c) 2011 Frane Šarić <frane.saric@fer.hr>
 
   Zadatak:
   Napišite program koji će pretraživanjem u širinu pronaći optimalno rješenje
   problema misionara i kanibala. Problem je opisan na sljedeći način: tri misionara
   i tri kanibala potrebno je jednim čamcem prevesti s jedne strane obale rijeke na
   drugu, pri čemu se niti u jednom trenutku na jednoj strani obale ne smije naći
   više kanibala nego misionara. Čamac može prevesti najviše dvije osobe i ne
   može ploviti prazan. Optimalno rješenje je ono s najmanjim brojem koraka.
   Program treba ispisati rješenje u obliku niza operatora i stanja koja dovode do
   ciljnog stanja.

   Primjer: http://www.learn4good.com/games/puzzle/boat.htm
*/
#include <vector>
#include <cstdio> // U C++-u treba koristiti cstdio umjesto stdio.h, itd.

using namespace std; // Da ne moramo pisati std::vector, itd.

struct stanje { // Pamtimo broj k. i m. na ovoj i suprotnoj obali.
    int k[2], m[2], o; // o = 0 je lijeva obala, o = 1 je desna.
    stanje(int k1, int m1, int k2, int m2, int o_) : o(o_) {
        k[0] = k1, k[1] = k2, m[0] = m1, m[1] = m2;
    }
    stanje prebaci(int nk, int nm) const { // Promijeni stranu rijeke.
        return stanje(k[1] + nk, m[1] + nm, k[0] - nk, m[0] - nm, 1 - o);
    }
    bool moze() const { // Je li dobro stanje (neće se nitko pojesti)?
        return (!m[0] || m[0] >= k[0]) && (!m[1] || m[1] >= k[1]);
    }
};

// Prvi parametar je referenca jer ga ne želimo kopirati svaki put kad se
// zove funkcija (poziv je "by value"), a const je jer ga ne mijenjamo.
void ispisi(const vector<pair<stanje, int> > &q, int qpos) {
    const stanje &s = q[qpos].first; // Trenutno stanje.
    if (q[qpos].second >= 0) { // Samo ako ovo nije početno stanje onda:
        ispisi(q, q[qpos].second); // ispiši korake do prethodnog stanja,
        const stanje &prev = q[q[qpos].second].first;
        printf("%5c--- K:%d M:%d ---%c\n", "< "[s.o], s.k[0] - prev.k[1],
                s.m[0] - prev.m[1], " >"[s.o]); // ispiši operaciju.
    }
    printf("K:%d M:%d ......... K:%d M:%d\n", // Ispiši trenutno stanje.
            s.k[s.o], s.m[s.o], s.k[1 - s.o], s.m[1 - s.o]);
}

int main() {
    static char visited[4][4][4][4][2]; // Koja stanja su posjećena?
    vector<pair<stanje, int> > q; // Pamtimo stanje i indeks prethodnog.
    q.push_back(make_pair(stanje(3, 3, 0, 0, 0), -1)); // Početno stanje.

    for (unsigned qstart = 0; qstart < q.size(); ++qstart) {
        stanje s = q[qstart].first; // Uzmi stanje s početka reda.
        if (s.o == 1 && s.k[0] + s.m[0] == 6) { // Našli smo rješenje?
            ispisi(q, qstart); // Ispiši ga i prekini traženje.
            break;
        }
        for (int k = 0; k <= s.k[0]; ++k) // Probaj staviti kanibale.
            for (int m = 0; m <= s.m[0]; ++m) { // Probaj staviti misionare.
                stanje n = s.prebaci(k, m); // Konstruiraj novo stanje.
                char &vis = visited[n.k[0]][n.k[1]][n.m[0]][n.m[1]][n.o];
                if (m + k > 0 && m + k <= 2 && n.moze() && !vis) {
                    q.push_back(make_pair(n, qstart)); // Dodaj u red.
                    vis = 1; // Označi stanje kao posjećeno.
                }
            }
    }
}
